lighthouse-core/test/gather/gatherers/js-usage-test.js

```diff
@@ -0,0 +1,97 @@
+
+describe('JsUsage gatherer', () => {
+  /**
+   * `scriptParsedEvent` mocks the `sourceMapURL` and `url` seen from the protocol.
+   * `map` mocks the (JSON) of the source maps that `Runtime.evaluate` returns.
+   * `resolvedSourceMapUrl` is used to assert that the SourceMaps gatherer is using the expected
+   *                        url to fetch the source map.
+   * `fetchError` mocks an error that happens in the page. Only fetch error message make sense.
+   * @param {{coverage: LH.Crdp.Profiler.ScriptCoverage[], scriptParsedEvents: LH.Crdp.Debugger.ScriptParsedEvent[]}} _
+   * @return {Promise<LH.Artifacts['JsUsage']>}
+   */
+  async function runJsUsage({coverage, scriptParsedEvents = []}) {
+    const onMock = createMockOnFn();
+    const sendCommandMock = createMockSendCommandFn()
+      .mockResponse('Profiler.enable', {})
+      .mockResponse('Profiler.disable', {})
+      .mockResponse('Profiler.startPreciseCoverage', {})
+      .mockResponse('Profiler.takePreciseCoverage', {result: coverage})
+      .mockResponse('Profiler.stopPreciseCoverage', {});
+
+    for (const scriptParsedEvent of scriptParsedEvents) {
+      onMock.mockEvent('protocolevent', {
+        method: 'Debugger.scriptParsed',
+        params: scriptParsedEvent,
+      });
+    }
+    const connectionStub = new Connection();
+    connectionStub.sendCommand = sendCommandMock;
+    connectionStub.on = onMock;
+
+    const driver = new Driver(connectionStub);
+
+    const gatherer = new JsUsage();
+    await gatherer.beforePass({driver});
+    jest.advanceTimersByTime(1);
+    return gatherer.afterPass({driver});
+  }
+
+  it('combines coverage data by url', async () => {
+    const coverage = [
+      {scriptId: '1', url: 'https://www.example.com'},
+      {scriptId: '2', url: 'https://www.example.com'},
+    ];
+    const artifact = await runJsUsage({coverage});
+    expect(artifact).toMatchInlineSnapshot(`
+      Object {
+        "https://www.example.com": Array [
+          Object {
+            "scriptId": "1",
+            "url": "https://www.example.com",
+          },
+          Object {
+            "scriptId": "2",
+            "url": "https://www.example.com",
+          },
+        ],
+      }
+    `);
+  });
+
+  it('ignore coverage data with empty url', async () => {
+    const coverage = [{scriptId: '1', url: ''}];
+    const artifact = await runJsUsage({coverage});
+    expect(artifact).toMatchInlineSnapshot(`Object {}`);
+  });
+
+  it('uses ScriptParsedEvent embedderName over coverage data url', async () => {
+    const coverage = [{scriptId: '1', url: 'LOL WHAT'}];
+    const scriptParsedEvents = [{scriptId: '1', embedderName: 'https://www.example.com'}];
+    const artifact = await runJsUsage({coverage, scriptParsedEvents});
+    expect(artifact).toMatchInlineSnapshot(`
+      Object {
+        "https://www.example.com": Array [
+          Object {
+            "scriptId": "1",
+            "url": "LOL WHAT",
```

Comment:
this is a somewhat confusing result, should we override here too and save this url under another property name?

thanks for these tests btw! :)

Reply:
We don't use that property anywhere. Maybe we should be mapping the `ScriptCoverage` instead of gathering the entire object. Do you want to modify this now or wait until it is needed?

Reply:
Gotcha, I worry about the wrong value sitting there and getting accidentally used in the future instead of the key and we *definitely* won't notice it in *that* review, heh

maybe just change the type of this artifact to `Omit<Crdp.Profiler.ScriptCoverage, 'url'>` so no one accidentally uses it?

Reply:
oops, actually no-unload-listeners uses it. easy fix tho.